Sources/SwiftDocC/Utility/CollectionChanges.swift

```diff
@@ -0,0 +1,219 @@
+            case .insert(let offset, _, _):
+                changes.insert(at: offset)
+            }
+        }
+        segments = changes.segments
+    }
+}
+
+/// A builder that applies collection differences to construct an array of ``Segment`` values.
+///
+/// - Important:
+/// Removals need to be applied in reverse order. All removals need to be applied before applying any insertions. Insertions need to be applied in order.
+private struct ChangeSegmentBuilder {
+    typealias Segment = CollectionChanges.Segment
+    
+    private(set) var segments: [Segment]
+    
+    private var insertStartIndex = 0
+    private var insertStartOffset = 0
+    
+    init(originalCount: Int) {
+        self.segments = [ Segment(kind: .common, count: originalCount) ]
+    }
+    
+    mutating func remove(at removalIndex: Int) {
+        // Removals are applied in reverse order. When the first removal is applied, the only segment is the 'original' count.
+        //
+        // Each removal can be either be at the start of the segment, middle of the segment, or end of the segment.
+        // - After removing from the start of the segment there can be no more removals (since those indices would be in ascending order).
+        // - After removing from the middle, the 'original' segment is split in two with a 'remove' segment in between.
+        //   Since the removal has to be at a lower index, it can only be applied to the split 'original' segment.
+        // - After removing from the end, the 'original' segment is made shorter and a new 'remove' segment is added after it.
+        //   Since the removal has to be at a lower index, it can only be applied to the shortened 'original' segment.
+        //
+        // This process repeats, meaning that every removal is always applied to the first segment.
+        let segment = segments[0]
+        precondition(segment.kind == .common && removalIndex < segment.count, """
+            The first segment should always be an 'original' segment (was \(segment.kind)) and (0 ..< \(segment.count)) should always contain the removal index (\(removalIndex)).
+            If it's not, then that's means that the remove operations wasn't performed in reverse order.
+            """)
+        
+        if removalIndex == 0 {
+            // Removing at the start of the segment
+            if segment.count == 1 {
+                segments.remove(at: 0)
+            } else {
+                segments[0].count -= 1
+            }
+            
+            if segments.isEmpty || segments[0].kind != .remove {
+                segments.insert(Segment(kind: .remove, count: 1), at: 0)
+            } else {
+                segments[0].count += 1
+                assert(segments[0].kind == .remove)
+            }
+        }
+        else if removalIndex == segment.count - 1 {
+            // Removing at end of segment
+            segments[0].count -= 1
+            assert(segments[0].count > 0, """
```

Comment:
I don't think this assert can hit? Wouldn't the control flow go inside the `removalIndex == 0` branch? Consider removing the assert

Reply:
You're right. This assertion was left in from an earlier version of the code. I've removed it.

Reply:
Fixed as part of https://github.com/apple/swift-docc/pull/420/commits/08346bf76824e7d7aef10aa65d6585c149236b31